### DISCUSSION TITLE
Redis connection pool sizes with Sidekiq Enterprise

### DISCUSSION THREAD

**Discussion by justin808** [sidekiq/sidekiq#5295] [General]
I’m switching a BIG project (sometimes 50+ concurrent P-L) servers from using deprecated `Redis.current` to https://github.com/mperham/connection_pool. Some servers use Puma. Others use Sidekiq Pro.

**What are the recommended defaults for the configuration options shown below?**

## Details
* We’re using one big Redis for both Sidekiq and caching items, like keeping an MRU of requests to playback for cache warming before releases.
* Because tracking items in the MRU is non-critical, we will have a short_timeout pool. 
* Plus, sometimes the load and connection requirements go way up when I’m running massive 24+ P-L cache warming jobs against Redis.
* Plus, using [Heroku Preboot](https://devcenter.heroku.com/articles/preboot), both the old and new servers can be talking to Redis, so 20 P-L prebooting can be like 40 servers!

## What should I be concerned about?
* What happens if I periodically run out of Redis connections since we don’t have a service like [PgBouncer for Heroku](https://devcenter.heroku.com/articles/best-practices-pgbouncer-configuration). (Why isn’t there such a service for Redis?)
* Any performance concerns? 
* Anything I should measure with [DataDog’s Redis and Sidekiq integration](https://www.datadoghq.com/blog/monitor-sidekiq-with-datadog/)?

Here are my guesses below.

Some Sidekiq concurrency, say 10, plus some Sidekiq swarm SIDEKIQ_COUNT, like 3, for the P-L servers.

## Configuration Guesses

default pool:
REDIS_DEFAULT_POOL_SIZE: Pool size (Default 1).
REDIS_DEFAULT_POOL_TIMEOUT: Pool timeout (Default 5 seconds).
REDIS_DEFAULT_CLIENT_TIMEOUT: Client timeout (Default 5 seconds).

short_timeout pool:
REDIS_SHORT_POOL_SIZE: Pool size (Default 1).
REDIS_SHORT_POOL_TIMEOUT: Pool timeout (Default 1 seconds).
REDIS_SHORT_CLIENT_TIMEOUT: Client timeout (Default 1 second).

**Comment by mperham** [comment]
I'd set all the timeouts to 1 sec. Don't set any size. Let the pool auto-size.

You might find that the default concurrency of 10 is too high and your P-L cores are constantly at 100%. It's ok to lower it, you don't want to overload your dyno.

If you run out of connections, the service will raise an error and you can adjust your dyno/process/concurrency count to lower your connection usage. Existing connections and jobs should continue processing.

**Reply by axsuul** [reply]
You mention in the wiki that `ConnectionPool` should be specified to have plenty of connections based on concurrency: https://github.com/mperham/sidekiq/wiki/Using-Redis#complete-control

Has this advice changed?

**Reply by mperham** [reply]
No. Don't specify a connection pool at all if possible, let Sidekiq build it. If you must supply a pool, ensure it is 3 larger than your concurrency.
